deploy/deploy_test.go

```diff
@@ -361,3 +389,412 @@ func TestOrdering(t *testing.T) {
 		})
 	}
 }
+
+type fileData struct {
+	Name     string // name of the file
+	Contents string // contents of the file
+}
+
+// initLocalFs initializes fs with some test files.
+func initLocalFs(ctx context.Context, fs afero.Fs) ([]*fileData, error) {
+	// The initial local filesystem.
+	local := []*fileData{
+		{"aaa", "aaa"},
+		{"bbb", "bbb"},
+		{"subdir/aaa", "subdir-aaa"},
+		{"subdir/nested/aaa", "subdir-nested-aaa"},
+		{"subdir2/bbb", "subdir2-bbb"},
+	}
+	if err := writeFiles(fs, local); err != nil {
+		return nil, err
+	}
+	return local, nil
+}
+
+// fsTest represents an (afero.FS, Go CDK blob.Bucket) against which end-to-end
+// tests can be run.
+type fsTest struct {
+	name   string
+	fs     afero.Fs
+	bucket *blob.Bucket
+}
+
+// initFsTests initializes a pair of tests for end-to-end test:
+// 1. An in-memory afero.Fs paired with an in-memory Go CDK bucket.
+// 2. A filesystem-based afero.Fs paired with an filesystem-based Go CDK bucket.
+// It returns the pair of tests and a cleanup function.
+func initFsTests() ([]*fsTest, func(), error) {
+	tmpfsdir, err := ioutil.TempDir("", "fs")
+	if err != nil {
+		return nil, nil, err
+	}
+	tmpbucketdir, err := ioutil.TempDir("", "bucket")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	memfs := afero.NewMemMapFs()
+	membucket := memblob.OpenBucket(nil)
```

Comment:
Since these tests write to memory and not to, say, Amason S3, you should rework the commit message title. These are not end-to-end tests; that would be "hugo && hugo deploy" with a real site and real buckets.

Reply:
Done.